lib/msf/core/post/file.rb

```diff
@@ -376,33 +389,104 @@ def write_file(file_name, data)
   #
-  # Platform-agnostic file append. Appends given object content to a remote file.
-  # Returns Boolean true if successful
+  # NOTE: *This is not binary-safe on Windows shell sessions!*
+  #
+  # @param file_name [String] Remote file name to write
+  # @param data [String] Contents to put in the file
+  # @param max_chunk [int] max size for the data chunk to write at a time
+  # @return [void]
+  def win_ansi_write_file(file_name, data, max_chunk = 5000)
+    start_index = 0
+    write_length =[max_chunk, data.length].min
+    session.shell_command_token("echo | set /p=\"#{data[0, write_length]}\"> \"#{file_name}\"")
+    if data.length > write_length
+      # just use append to finish the rest
+      win_ansi_append_file(file_name, data[write_length, data.length], max_chunk)
+    end
+  end
+
+  # Windows ansi file append for shell sessions. Writes given object content to a remote file.
   #
   # NOTE: *This is not binary-safe on Windows shell sessions!*
   #
   # @param file_name [String] Remote file name to write
   # @param data [String] Contents to put in the file
+  # @param max_chunk [int] max size for the data chunk to write at a time
+  # @return [void]
+  def win_ansi_append_file(file_name, data, max_chunk = 5000)
+    start_index = 0
+    write_length =[max_chunk, data.length].min
+    while start_index < data.length
+      session.shell_command_token("<nul set /p=\"#{data[start_index, write_length]}\" >> \"#{file_name}\"")
+      start_index = start_index + write_length
+      write_length = [max_chunk, data.length - start_index].min
+    end
+  end
+
+  # Windows binary file write for shell sessions. Writes given object content to a remote file.
+  #
+  # @param file_name [String] Remote file name to write
+  # @param data [String] Contents to put in the file
+  # @param max_chunk [int] max size for the data chunk to write at a time
+  # @return [void]
+  def win_bin_write_file(file_name, data, max_chunk = 5000)
+    b64_data = Base64.strict_encode64(data)
+    b64_filename = "#{file_name}.b64"
+    win_ansi_write_file(b64_filename, b64_data, max_chunk)
+    cmd_exec("certutil -decode #{b64_filename} #{file_name}")
+    file_rm(b64_filename)
+  end
+
+  # Windows binary file append for shell sessions. Appends given object content to a remote file.
+  #
+  # @param file_name [String] Remote file name to write
+  # @param data [String] Contents to put in the file
+  # @param max_chunk [int] max size for the data chunk to write at a time
   # @return [void]
+  def win_bin_append_file(file_name, data, max_chunk = 5000)
+    b64_data = Base64.strict_encode64(data)
+    b64_filename = "#{file_name}.b64"
+    tmp_filename = "#{file_name}.tmp"
+    win_ansi_write_file(b64_filename, b64_data, max_chunk)
+    cmd_exec("certutil -decode #{b64_filename} #{tmp_filename}")
+    cmd_exec("copy /b #{file_name}+#{tmp_filename} #{file_name}")
+    file_rm(b64_filename)
```

Comment:
Do we need to ensure we're always cleaning up these files incase something goes wrong?